src/fs/pfs_filesystem_impl.cpp

```diff
@@ -0,0 +1,505 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * File Created: 2022-06-01
+ * Author: xuchaojie
+ */
+
+#include "src/fs/pfs_filesystem_impl.h"
+
+#include <glog/logging.h>
+#include <sys/vfs.h>
+#include <sys/utsname.h>
+#include <linux/version.h>
+#include <dirent.h>
+#include <brpc/server.h>
+#include <butil/iobuf.h>
+
+#include <pfs_api.h>
+#include <sys/uio.h>
+
+#include <string>
+#include <vector>
+
+#include "src/common/string_util.h"
+
+
+#define NVME_ALIGN 8
+#define SECTOR_SIZE 512
+
+namespace curve {
+namespace fs {
+
+std::string DumpIOVec(const struct iovec *vec, size_t nvec) {
+    std::stringstream ss;
+    ss << "{";
+    for (size_t i = 0; i < nvec; ++i) {
+        ss << "[" << uintptr_t(vec[i].iov_base) << ","
+           << vec[i].iov_len << "]";
+    }
+    ss << "}";
+    return ss.str();
+}
+
+ssize_t preadv_dispatch(int fd, const struct iovec *vector,
+                        int nvec, off_t offset) {
+    if (PFS_FD_ISVALID(fd)) {
+        int ret = pfs_preadv_dma(fd, vector, nvec, offset);
+        if (ret < 0) {
+            LOG(ERROR) << "preadv_dispatch failed, ret: " << ret
+                       << ", fd: " << fd
+                       << ", nvec: " << nvec
+                       << ", offset: " << offset
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return preadv(fd, vector, nvec, offset);
+}
+
+ssize_t pwritev_dispatch(int fd, const struct iovec *vector,
+                         int nvec, off_t offset) {
+    if (PFS_FD_ISVALID(fd)) {
+        // TODO(xuchaojie): use pfs_pwritev_dma in future
+        int ret = pfs_pwritev(fd, vector, nvec, offset);
+        if (ret < 0) {
+            LOG(ERROR) << "pwritev_dispatch failed, ret: " << ret
+                       << ", fd: " << fd
+                       << ", nvec: " << nvec
+                       << ", offset: " << offset
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return pwritev(fd, vector, nvec, offset);
+}
+
+ssize_t readv_dispatch(int fd, const struct iovec *vector, int nvec) {
+    if (PFS_FD_ISVALID(fd)) {
+        int ret = pfs_readv_dma(fd, vector, nvec);
+        if (ret < 0) {
+            LOG(ERROR) << "readv_dispatch failed, ret: " << ret
+                       << ", fd: " << fd
+                       << ", nvec: " << nvec
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return readv(fd, vector, nvec);
+}
+
+ssize_t writev_dispatch(int fd, const struct iovec *vector, int nvec) {
```

Comment:
ditto

Reply:
What's the diffrence between pwritev_dispatch and writev_dispatch.

Reply:
pwritev_dispatch's parameter has offset feild